runtime/compiler/optimizer/EscapeAnalysis.cpp

```diff
@@ -6698,6 +6698,149 @@ bool TR_EscapeAnalysis::tryToZeroInitializeUsingArrayset(Candidate* candidate, T
+      ttNode->setAndIncChild(0, fieldValueNode);
+
+      fieldValueTreeTopCursor = TR::TreeTop::create(comp(), fieldValueTreeTopCursor, ttNode);
+
+      // generate store to the field
+      const TR::TypeLayoutEntry& fieldEntry = typeLayout->entry(i - 1);
+      TR::SymbolReference* symref = comp()->getSymRefTab()->findOrFabricateShadowSymbol(valueClass,
+                                                                       fieldEntry._datatype,
+                                                                       fieldEntry._offset,
+                                                                       fieldEntry._isVolatile,
+                                                                       fieldEntry._isPrivate,
+                                                                       fieldEntry._isFinal,
+                                                                       fieldEntry._fieldname,
+                                                                       fieldEntry._typeSignature
+                                                                       );
+
+      int32_t fieldSize = TR::Symbol::convertTypeToSize(fieldEntry._datatype);
+
+      if (fieldEntry._datatype.getDataType() == TR::Address && comp()->useCompressedPointers())
+         {
+         fieldSize = TR::Compiler->om.sizeofReferenceField();
+         }
+
+      TR::Node *completeFieldStoreNode = NULL;
+
+      FieldInfo &field = candidate->findOrSetFieldInfo(NULL, symref, fieldEntry._offset, fieldSize, fieldEntry._datatype, this);
+
+      if (candidate->isContiguousAllocation())
+         {
+         field._symRef = symref;
+         field._vectorElem = 0;
+
+         TR::DataType fieldDataType = fieldValueNode->getDataType();
+         const bool useWriteBarrier = ( (fieldDataType == TR::Address)
+                                        && (TR::Compiler->om.writeBarrierType() != gc_modron_wrtbar_none) )
+                                      || comp()->getOption(TR_EnableFieldWatch);
+         TR::Node *storeNode;
+
+         if (useWriteBarrier)
+            {
+            const TR::ILOpCodes storeOpCode = comp()->il.opCodeForIndirectWriteBarrier(fieldValueNode->getDataType());
+            storeNode = TR::Node::createWithSymRef(storeOpCode, 3, 3, candidate->_node, fieldValueNode, candidate->_node, symref);
+            }
+         else
+            {
+            const TR::ILOpCodes storeOpCode = comp()->il.opCodeForIndirectStore(fieldValueNode->getDataType());
+            storeNode = TR::Node::createWithSymRef(storeOpCode, 2, 2, candidate->_node, fieldValueNode, symref);
+            }
+
+         // if storing a ref, make sure it is compressed
+         if (comp()->useCompressedPointers() && fieldValueNode->getDataType() == TR::Address)
+            {
+            completeFieldStoreNode = TR::Node::createCompressedRefsAnchor(storeNode);
+            }
+         else
+            {
+            completeFieldStoreNode = storeNode;
+            }
```

Comment:
The question here is just to help me understand the code better: With my understanding of how [non contiguous](https://github.com/eclipse-openj9/openj9/pull/17594#discussion_r1235793546) doesn't need to consider compressed ref, contiguous stack allocation also stores the field value into an offset of a local temporary (no heap access). Why would compressed ref be needed in contiguous case?

Reply:
In general a contiguously stack allocated object might be accessed at a point in the code that might also be able to access a heap allocated object.  Consider a case like the following, where class `A` has a reference field `r`.  After the `if` construct, `a.r` might be referring to a field in the object created at the line marked `Obj 1` - which must be allocated on the heap, because it escapes - or a field in the object created at the line marked `Obj 2` - which might be stack allocated, because it doesn't escape.  So the offsets and sizes of the shadow symbols used to access the fields in the contiguously stack allocated object have to be the same as those used to access a heap allocated instance of the same class.

```
static A s = new A();  // Obj 1
A a = new A();   // Obj 2
if (some-condition) {
    a = s;
}
Object o = a.r;   // Obj 1 or 2?
```